File: education/education/page/schedule_dashboard/schedule_dashboard.js
frappe.pages['schedule-dashboard'].on_page_load = function (wrapper) {
  const page = frappe.ui.make_app_page({
    parent: wrapper,
    title: __('Schedule Dashboard'),
    single_column: true,
  })

  wrapper.schedule_dashboard = new ScheduleDashboard(page, wrapper)
}

frappe.pages['schedule-dashboard'].on_page_show = function (wrapper) {
  if (wrapper.schedule_dashboard) {
    wrapper.schedule_dashboard.on_show()
  }
}

// ─── helpers ───────────────────────────────────────────────────────────────

function fmt12(time_str) {
  if (!time_str) return ''
  const parts = String(time_str).split(':')
  let h = parseInt(parts[0], 10)
  const m = parts[1] || '00'
  const ampm = h >= 12 ? 'PM' : 'AM'
  if (h === 0) h = 12
  else if (h > 12) h -= 12
  return `${h}:${m} ${ampm}`
}

const COLOR_BG = {
  blue: '#EDF6FD',
  green: '#E4F5E9',
  red: '#FFF0F0',
  orange: '#FFF1E7',
  yellow: '#FFF7D3',
  teal: '#E6F7F4',
  violet: '#F5F2FF',
  cyan: '#E0F8FF',
  amber: '#FCF3CF',
  pink: '#FEEEF8',
  purple: '#F9F0FF',
}

// ─── Page class ────────────────────────────────────────────────────────────

class ScheduleDashboard {
  constructor(page, wrapper) {
    this.page = page
    this.wrapper = wrapper
    this._setup_fields()
    this._setup_body()
    this._toggle_fields()
    this._show_prompt()
  }

  // ── setup ──────────────────────────────────────────────────────────────

  _setup_fields() {
    // All fields are added without hidden:1 so they exist in the DOM.
    // Visibility is controlled manually via _toggle_fields().

    this.f_view = this.page.add_field({
      label: __('View'),
      fieldtype: 'Select',
      fieldname: 'view',
      default: 'student_group_now',
      options: [
        { label: __('Who is Teaching Now (Student Group)'), value: 'student_group_now' },
        { label: __('Instructor Daily Schedule'), value: 'instructor_daily' },
        { label: __('Student Group Timetable'), value: 'student_group_timetable' },
      ],
      change: () => {
        this.f_instructor.set_value('')
        this._toggle_fields()
        this._show_prompt()
      },
    })

    this.f_student_group = this.page.add_field({
      label: __('Student Group'),
      fieldtype: 'Link',
      fieldname: 'student_group',
      options: 'Student Group',
      change: () => this.load(),
    })

    this.f_instructor = this.page.add_field({
      label: __('Instructor'),
      fieldtype: 'Link',
      fieldname: 'instructor',
      options: 'Instructor',
      get_query: () => {
        const date = this.f_date ? this.f_date.get_value() : frappe.datetime.get_today()
        return {
          query:
            'education.education.page.schedule_dashboard.schedule_dashboard.get_instructors_query',
          filters: { date: date || frappe.datetime.get_today() },
        }
      },
      change: () => this.load(),
    })

    this.f_date = this.page.add_field({
      label: __('Date'),
      fieldtype: 'Date',
      fieldname: 'date',
      default: frappe.datetime.get_today(),
      change: () => {
        // Clear instructor so user re-picks from the updated filtered list
        if (this._view() === 'instructor_daily') {
          this.f_instructor.set_value('')
          this._show_prompt()
        } else {
          this.load()
        }
      },
    })
  }

  _setup_body() {
    this.$body = $('<div class="sd-results" style="padding:20px 15px;min-height:200px;"></div>')
    // Use wrapper directly — guaranteed to contain .layout-main-section
    $(this.wrapper).find('.layout-main-section').append(this.$body)
  }

  // ── visibility ─────────────────────────────────────────────────────────

  _view() {
    return this.f_view.get_value() || 'student_group_now'
  }

  _toggle_fields() {
    const v = this._view()
    this.f_student_group.$wrapper.toggle(v !== 'instructor_daily')
    this.f_instructor.$wrapper.toggle(v === 'instructor_daily')
    this.f_date.$wrapper.toggle(v !== 'student_group_now')
  }

  // ── public ─────────────────────────────────────────────────────────────

  on_show() {
    // Refresh data when navigating back to this page
    this.load()
  }

  load() {
    const v = this._view()
    if (v === 'student_group_now') this._load_sg_now()
    else if (v === 'instructor_daily') this._load_instructor_daily()
    else if (v === 'student_group_timetable') this._load_sg_timetable()
  }

  // ── helpers ────────────────────────────────────────────────────────────

  _show_prompt() {
    const v = this._view()
    if (v === 'student_group_now') {
      this.$body.html(
        `<p class="text-muted" style="margin-top:20px;">${__('Select a Student Group to see who is teaching right now.')}</p>`
      )
    } else if (v === 'instructor_daily') {
      this.$body.html(
        `<p class="text-muted" style="margin-top:20px;">${__('Select a date then choose an Instructor to see their daily schedule.')}</p>`
      )
    } else {
      this.$body.html(
        `<p class="text-muted" style="margin-top:20px;">${__('Select a Student Group to see the timetable.')}</p>`
      )
    }
  }

  _loading() {
    this.$body.html(
      `<div style="padding:30px;text-align:center;">
        <div class="text-muted">${frappe.utils.icon('loading', 'lg')} ${__('Loading...')}</div>
      </div>`
    )
  }

  _empty(msg) {
    this.$body.html(
      `<div style="padding:40px;text-align:center;">
        <p class="text-muted">${msg}</p>
      </div>`
    )
  }

  // ── view loaders ───────────────────────────────────────────────────────

  _load_sg_now() {
    const sg = this.f_student_group.get_value()
    if (!sg) {
      this._show_prompt()
      return
    }

    this._loading()

    frappe.call({
      method:
        'education.education.page.schedule_dashboard.schedule_dashboard.get_student_group_current_instructor',
      args: { student_group: sg },
      callback: (r) => {
        const data = r.message || []
        if (!data.length) {
          this._empty(
            `${__('No class is currently in session for')} <strong>${frappe.utils.escape_html(sg)}</strong>`
          )
          return
        }

        let html = `<h5 style="margin-bottom:16px;">${__('Currently in session')} — <strong>${frappe.utils.escape_html(sg)}</strong></h5><div class="row">`

        for (const s of data) {
          html += `
            <div class="col-sm-6 col-md-4" style="margin-bottom:14px;">
              <div style="background:${COLOR_BG[s.class_schedule_color] || '#fff'};border-left:4px solid var(--primary);padding:14px 16px;border-radius:6px;box-shadow:0 1px 4px rgba(0,0,0,.06);">
                <h5 style="margin:0 0 8px;">${frappe.utils.escape_html(s.course)}</h5>
                <p style="margin:0 0 4px;"><strong>${__('Instructor')}:</strong> ${frappe.utils.escape_html(s.instructor_name || s.instructor)}</p>
                <p style="margin:0 0 4px;"><strong>${__('Time')}:</strong> ${fmt12(s.from_time)} – ${fmt12(s.to_time)}</p>
                <p style="margin:0;"><strong>${__('Room')}:</strong> ${frappe.utils.escape_html(s.room || '—')}</p>
              </div>
            </div>`
        }

        html += '</div>'
        this.$body.html(html)
      },
      error: () => {
        this._empty(__('Error loading schedule. Please try again.'))
      },
    })
  }

  _load_instructor_daily() {
    const instructor = this.f_instructor.get_value()
    const date = this.f_date.get_value() || frappe.datetime.get_today()

    if (!instructor) {
      this._show_prompt()
      return
    }

    this._loading()

    frappe.call({
      method:
        'education.education.page.schedule_dashboard.schedule_dashboard.get_instructor_daily_schedule',
      args: { instructor, date },
      callback: (r) => {
        const data = r.message || []
        const display_date = frappe.datetime.str_to_user(date)

        if (!data.length) {
          this._empty(
            `${__('No classes scheduled for')} <strong>${frappe.utils.escape_html(data[0] ? data[0].instructor_name : instructor)}</strong> ${__('on')} ${display_date}`
          )
          return
        }

        const name_display = frappe.utils.escape_html(data[0].instructor_name || instructor)

        let html = `
          <h5 style="margin-bottom:14px;">${__('Schedule for')} <strong>${name_display}</strong> — ${display_date}</h5>
          <table class="table table-bordered" style="font-size:13px;">
            <thead class="grid-heading-row">
              <tr>
                <th>${__('Time')}</th>
                <th>${__('Subject')}</th>
                <th>${__('Student Group')}</th>
                <th>${__('Room')}</th>
                <th>${__('Status')}</th>
              </tr>
            </thead>
            <tbody>`

        for (const s of data) {
          const row_style = s.is_ongoing ? 'background:#d9f5e4;font-weight:600;' : ''
          const badge = s.is_ongoing
            ? `<span class="indicator-pill green nowrap">${__('Ongoing')}</span>`
            : ''

          html += `
            <tr style="${row_style}">
              <td class="nowrap">${fmt12(s.from_time)} – ${fmt12(s.to_time)}</td>
              <td><a href="/app/course-schedule/${s.name}">${frappe.utils.escape_html(s.course)}</a></td>
              <td><a href="/app/student-group/${encodeURIComponent(s.student_group)}">${frappe.utils.escape_html(s.student_group)}</a></td>
              <td>${frappe.utils.escape_html(s.room || '—')}</td>
              <td>${badge}</td>
            </tr>`
        }

        html += '</tbody></table>'
        this.$body.html(html)
      },
      error: () => {
        this._empty(__('Error loading schedule. Please try again.'))
      },
    })
  }

  _load_sg_timetable() {
    const sg = this.f_student_group.get_value()
    const date = this.f_date.get_value() || frappe.datetime.get_today()

    if (!sg) {
      this._show_prompt()
      return
    }

    this._loading()

    frappe.call({
      method:
        'education.education.page.schedule_dashboard.schedule_dashboard.get_student_group_timetable',
      args: { student_group: sg, date },
      callback: (r) => {
        const result = r.message || {}
        const periods = result.periods || []
        const display_date = frappe.datetime.str_to_user(date)

        if (!periods.length) {
          const prog = frappe.utils.escape_html(result.program || '')
          this._empty(
            `${__('No Period Time Blocks found for program')} <strong>${prog}</strong>. ${__('Create Period Time Block records for this program first.')}`
          )
          return
        }

        let html = `
          <h5 style="margin-bottom:14px;">${__('Timetable for')} <strong>${frappe.utils.escape_html(sg)}</strong> — ${display_date}</h5>
          <table class="table table-bordered" style="font-size:13px;">
            <thead class="grid-heading-row">
              <tr>
                <th style="width:130px;">${__('Period')}</th>
                <th style="width:155px;">${__('Time')}</th>
                <th>${__('Subject')}</th>
                <th>${__('Instructor')}</th>
                <th>${__('Room')}</th>
              </tr>
            </thead>
            <tbody>`

        for (const p of periods) {
          const row_style = p.is_ongoing ? 'background:#d9f5e4;font-weight:600;' : ''
          const badge = p.is_ongoing
            ? ` <span class="indicator-pill green nowrap">${__('Now')}</span>`
            : ''

          const label = frappe.utils.escape_html(
            p.period_label || `${__('Period')} ${p.period_number}`
          )

          const subject_cell = p.schedule_name
            ? `<a href="/app/course-schedule/${p.schedule_name}">${frappe.utils.escape_html(p.course)}</a>`
            : `<span class="text-muted">—</span>`

          const instructor_cell = p.instructor
            ? `<a href="/app/instructor/${encodeURIComponent(p.instructor)}">${frappe.utils.escape_html(p.instructor_name || p.instructor)}</a>`
            : `<span class="text-muted">—</span>`

          html += `
            <tr style="${row_style}">
              <td>${label}${badge}</td>
              <td class="nowrap">${fmt12(p.from_time)} – ${fmt12(p.to_time)}</td>
              <td>${subject_cell}</td>
              <td>${instructor_cell}</td>
              <td>${frappe.utils.escape_html(p.room || '—')}</td>
            </tr>`
        }

        html += '</tbody></table>'
        this.$body.html(html)
      },
      error: () => {
        this._empty(__('Error loading timetable. Please try again.'))
      },
    })
  }
}

File: education/education/page/schedule_dashboard/schedule_dashboard.py
# Copyright (c) 2026, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import cint, getdate, now_datetime, nowdate


def _fmt_time(t):
	"""Convert a DB TIME value (timedelta or string) to a consistent HH:MM:SS string."""
	if t is None:
		return ""
	from frappe.utils import get_time

	t_obj = get_time(t)
	if t_obj:
		return t_obj.strftime("%H:%M:%S")
	return str(t) if t else ""


def _is_ongoing(from_time, to_time):
	"""Return True if the current server time falls within [from_time, to_time]."""
	if not from_time or not to_time:
		return False
	from frappe.utils import get_time

	now_t = now_datetime().time()
	try:
		from_t = get_time(from_time)
		to_t = get_time(to_time)
		return from_t <= now_t <= to_t
	except Exception:
		return False


@frappe.whitelist()
def get_student_group_current_instructor(student_group):
	"""Return classes currently in session for a student group (today, current time)."""
	today = nowdate()
	now = now_datetime().strftime("%H:%M:%S")

	rows = frappe.db.sql(
		"""
		SELECT
			cs.name, cs.course, cs.instructor, cs.instructor_name,
			cs.from_time, cs.to_time, cs.room, cs.class_schedule_color
		FROM `tabCourse Schedule` cs
		WHERE cs.student_group = %s
			AND cs.schedule_date = %s
			AND cs.from_time <= %s
			AND cs.to_time >= %s
		ORDER BY cs.from_time
		""",
		(student_group, today, now, now),
		as_dict=True,
	)

	for r in rows:
		r["from_time"] = _fmt_time(r["from_time"])
		r["to_time"] = _fmt_time(r["to_time"])

	return rows


@frappe.whitelist()
def get_instructor_daily_schedule(instructor, date=None):
	"""Return all Course Schedules for an instructor on a given date."""
	if not date:
		date = nowdate()

	rows = frappe.db.sql(
		"""
		SELECT
			cs.name, cs.course, cs.student_group, cs.instructor_name,
			cs.from_time, cs.to_time, cs.room, cs.schedule_date,
			cs.class_schedule_color, cs.program
		FROM `tabCourse Schedule` cs
		WHERE cs.instructor = %s
			AND cs.schedule_date = %s
		ORDER BY cs.from_time
		""",
		(instructor, date),
		as_dict=True,
	)

	is_today = getdate(date) == getdate(nowdate())

	for r in rows:
		from_raw = r["from_time"]
		to_raw = r["to_time"]
		r["from_time"] = _fmt_time(from_raw)
		r["to_time"] = _fmt_time(to_raw)
		r["is_ongoing"] = is_today and _is_ongoing(from_raw, to_raw)

	return rows


@frappe.whitelist()
def get_student_group_timetable(student_group, start_date=None, end_date=None, date=None):
	"""Return period-by-period timetable for a student group over a date range.

	Accepts either a single ``date`` (backward-compat) or a ``start_date``/
	``end_date`` range.  Dates that have no Course Schedule records are omitted
	from the results so the UI never renders an empty table.
	"""
	from datetime import timedelta

	# Backward-compat: single date via old parameter name
	if date and not start_date:
		start_date = date
	if not start_date:
		start_date = nowdate()
	if not end_date:
		end_date = start_date

	program = frappe.db.get_value("Student Group", student_group, "program")
	if not program:
		return {"results": [], "program": None}

	periods = frappe.get_all(
		"Period Time Block",
		filters={"program": program},
		fields=["period_number", "period_label", "from_time", "to_time"],
		order_by="period_number asc",
	)

	if not periods:
		return {"results": [], "program": program}

	# Pre-format period times once so inner-loop comparisons are cheap
	for p in periods:
		p["from_time_fmt"] = _fmt_time(p["from_time"])
		p["to_time_fmt"] = _fmt_time(p["to_time"])

	start_dt = getdate(start_date)
	end_dt = getdate(end_date)
	today = getdate(nowdate())

	_DAY_NAMES = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday"]

	results = []
	current = start_dt

	while current <= end_dt:
		current_str = str(current)

		schedules = frappe.db.sql(
			"""
			SELECT
				cs.name, cs.course, cs.instructor, cs.instructor_name,
				cs.from_time, cs.to_time, cs.room, cs.class_schedule_color
			FROM `tabCourse Schedule` cs
			WHERE cs.student_group = %s
			  AND cs.schedule_date = %s
			ORDER BY cs.from_time
			""",
			(student_group, current_str),
			as_dict=True,
		)

		# Skip dates that have no scheduled classes
		if schedules:
			is_today = current == today

			# Build a lookup keyed by the schedule's from_time (HH:MM:SS).
			# Matching by from_time alone is more robust than requiring an exact
			# from_time AND to_time match, in case time values differ slightly.
			sched_by_from = {}
			for s in schedules:
				ft = _fmt_time(s["from_time"])
				if ft not in sched_by_from:
					sched_by_from[ft] = s

			period_data = []
			for p in periods:
				matched = sched_by_from.get(p["from_time_fmt"])
				period_data.append(
					{
						"period_number": p["period_number"],
						"period_label": p["period_label"],
						"from_time": p["from_time_fmt"],
						"to_time": p["to_time_fmt"],
						"course": matched["course"] if matched else None,
						"instructor": matched["instructor"] if matched else None,
						"instructor_name": matched["instructor_name"] if matched else None,
						"room": matched["room"] if matched else None,
						"schedule_name": matched["name"] if matched else None,
						"color": matched["class_schedule_color"] if matched else None,
						"is_ongoing": is_today and _is_ongoing(p["from_time"], p["to_time"]),
					}
				)

			results.append(
				{
					"date": current_str,
					"day_of_week": _DAY_NAMES[current.weekday()],
					"periods": period_data,
				}
			)

		current = current + timedelta(days=1)

	return {"results": results, "program": program}


@frappe.whitelist()
def get_instructors_query(doctype, txt, searchfield, start, page_len, filters=None):
	"""Search query for Instructor link field filtered to instructors scheduled on a given date."""
	import json

	if isinstance(filters, str):
		try:
			filters = json.loads(filters)
		except Exception:
			filters = {}

	filters = filters or {}
	date = filters.get("date") or nowdate()
	txt = f"%{txt or ''}%"

	return frappe.db.sql(
		"""
		SELECT i.name, i.instructor_name
		FROM `tabInstructor` i
		WHERE i.name IN (
			SELECT DISTINCT cs.instructor
			FROM `tabCourse Schedule` cs
			WHERE cs.schedule_date = %(date)s
		)
		AND (i.name LIKE %(txt)s OR i.instructor_name LIKE %(txt)s)
		ORDER BY i.instructor_name
		LIMIT %(start)s, %(page_len)s
		""",
		{
			"date": date,
			"txt": txt,
			"start": cint(start),
			"page_len": cint(page_len),
		},
	)
